File: src/lib/sim/earnings.ts
import type { AssetSignal, SimAsset } from './strategies';

export type EarningsStatus = 'CLEAR' | 'BLOCKED' | 'UNKNOWN';

export interface EarningsRisk {
  symbol: string;
  status: EarningsStatus;
  nextDate: string | null;
  businessDaysUntil: number | null;
  source: 'NASDAQ';
  note: string;
}

const LOOKAHEAD_DAYS = 8;
const EXIT_BUSINESS_DAYS = 2;
const CACHE_TTL_MS = 6 * 60 * 60 * 1000;

let cache: { fetchedAt: number; risks: Map<string, EarningsRisk> } | null = null;

function isoDate(date: Date): string {
  return date.toISOString().slice(0, 10);
}

function businessDaysBetween(start: Date, end: Date): number {
  let days = 0;
  const cursor = new Date(start);
  cursor.setUTCHours(0, 0, 0, 0);
  const target = new Date(end);
  target.setUTCHours(0, 0, 0, 0);
  while (cursor < target) {
    cursor.setUTCDate(cursor.getUTCDate() + 1);
    const weekday = cursor.getUTCDay();
    if (weekday !== 0 && weekday !== 6) days += 1;
  }
  return days;
}

async function fetchNasdaqDate(date: string): Promise<Set<string>> {
  const response = await fetch(`https://api.nasdaq.com/api/calendar/earnings?date=${date}`, {
    headers: {
      accept: 'application/json, text/plain, */*',
      origin: 'https://www.nasdaq.com',
      referer: 'https://www.nasdaq.com/market-activity/earnings',
      'user-agent': 'Mozilla/5.0 (compatible; ATLAS-Macro-Terminal/1.0)',
    },
    signal: AbortSignal.timeout(8_000),
    cache: 'no-store',
  });
  if (!response.ok) throw new Error(`Nasdaq earnings HTTP ${response.status}`);
  const payload = await response.json() as {
    data?: { rows?: Array<{ symbol?: string }> | null };
  };
  return new Set((payload.data?.rows ?? []).flatMap((row) => row.symbol ? [row.symbol] : []));
}

export async function getEarningsRisks(
  universe: SimAsset[],
  now = new Date()
): Promise<Map<string, EarningsRisk>> {
  const stocks = universe.filter((asset) => asset.thematicEquity);
  if (cache && Date.now() - cache.fetchedAt < CACHE_TTL_MS) return cache.risks;

  const dates = Array.from({ length: LOOKAHEAD_DAYS + 1 }, (_, offset) => {
    const date = new Date(now);
    date.setUTCDate(date.getUTCDate() + offset);
    return isoDate(date);
  });
  const results = await Promise.allSettled(
    dates.map(async (date) => ({ date, symbols: await fetchNasdaqDate(date) }))
  );
  const failed = results.some((result) => result.status === 'rejected');
  const eventBySymbol = new Map<string, string>();
  for (const result of results) {
    if (result.status !== 'fulfilled') continue;
    for (const symbol of result.value.symbols) {
      if (!eventBySymbol.has(symbol)) eventBySymbol.set(symbol, result.value.date);
    }
  }

  const risks = new Map<string, EarningsRisk>();
  for (const asset of stocks) {
    const nextDate = eventBySymbol.get(asset.symbol) ?? null;
    if (failed) {
      risks.set(asset.symbol, {
        symbol: asset.symbol,
        status: 'UNKNOWN',
        nextDate,
        businessDaysUntil: nextDate ? businessDaysBetween(now, new Date(`${nextDate}T12:00:00Z`)) : null,
        source: 'NASDAQ',
        note: 'Calendar incomplete; new exposure is blocked conservatively',
      });
      continue;
    }
    const businessDaysUntil = nextDate
      ? businessDaysBetween(now, new Date(`${nextDate}T12:00:00Z`))
      : null;
    const blocked = businessDaysUntil != null && businessDaysUntil <= EXIT_BUSINESS_DAYS;
    risks.set(asset.symbol, {
      symbol: asset.symbol,
      status: blocked ? 'BLOCKED' : 'CLEAR',
      nextDate,
      businessDaysUntil,
      source: 'NASDAQ',
      note: blocked
        ? `Earnings window: ${businessDaysUntil} business day(s) until expected report`
        : nextDate
          ? `Expected report in ${businessDaysUntil} business day(s)`
          : `No expected report in the next ${LOOKAHEAD_DAYS} calendar days`,
    });
  }

  cache = { fetchedAt: Date.now(), risks };
  return risks;
}

export function applyEarningsPolicy(
  signals: AssetSignal[],
  risks: Map<string, EarningsRisk>
): AssetSignal[] {
  return signals.map((signal) => {
    if (!signal.thematicEquity) return signal;
    const risk = risks.get(signal.symbol);
    if (!risk || risk.status === 'CLEAR') return signal;
    return {
      ...signal,
      totalWeight: 0,
      weights: { tsmom: 0, carry: 0, macro: 0 },
      rationale: [
        ...signal.rationale,
        risk.status === 'BLOCKED'
          ? `${risk.note}; position forced flat before earnings`
          : `${risk.note}; target forced flat`,
      ],
    };
  });
}

File: services/model-engine/earnings.py
from __future__ import annotations

import urllib.request
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional, Set

from market import get_json

LOOKAHEAD_DAYS = 8
EXIT_BUSINESS_DAYS = 2


def iso_date(value: datetime) -> str:
    return value.date().isoformat()


def business_days_between(start: datetime, end: datetime) -> int:
    days = 0
    cursor = start.replace(hour=0, minute=0, second=0, microsecond=0)
    target = end.replace(hour=0, minute=0, second=0, microsecond=0)
    while cursor < target:
        cursor += timedelta(days=1)
        if cursor.weekday() < 5:
            days += 1
    return days


def fetch_nasdaq_date(day: str) -> Set[str]:
    url = f"https://api.nasdaq.com/api/calendar/earnings?date={day}"
    req = urllib.request.Request(
        url,
        headers={
            "Accept": "application/json, text/plain, */*",
            "Origin": "https://www.nasdaq.com",
            "Referer": "https://www.nasdaq.com/market-activity/earnings",
            "User-Agent": "Mozilla/5.0 (compatible; ATLAS-Macro-Terminal/1.0)",
        },
    )
    with urllib.request.urlopen(req, timeout=8) as response:
        import json

        payload = json.loads(response.read().decode("utf-8"))
    rows = ((payload.get("data") or {}).get("rows") or [])
    return {str(row["symbol"]).upper() for row in rows if row.get("symbol")}


def earnings_risks(symbols: List[str], now: Optional[datetime] = None) -> Dict[str, Dict[str, Any]]:
    now = now or datetime.now(timezone.utc)
    normalized = [symbol.strip().upper() for symbol in symbols if symbol.strip()]
    dates = [iso_date(now + timedelta(days=offset)) for offset in range(LOOKAHEAD_DAYS + 1)]
    event_by_symbol: Dict[str, str] = {}
    failed = False

    for day in dates:
        try:
            symbols_on_day = fetch_nasdaq_date(day)
            for symbol in symbols_on_day:
                event_by_symbol.setdefault(symbol, day)
        except Exception:
            failed = True

    risks: Dict[str, Dict[str, Any]] = {}
    for symbol in normalized:
        next_date = event_by_symbol.get(symbol)
        business_days_until = (
            business_days_between(now, datetime.fromisoformat(f"{next_date}T12:00:00+00:00"))
            if next_date
            else None
        )
        if failed:
            status = "UNKNOWN"
            note = "Calendar incomplete; new exposure is blocked conservatively"
        else:
            blocked = business_days_until is not None and business_days_until <= EXIT_BUSINESS_DAYS
            status = "BLOCKED" if blocked else "CLEAR"
            if blocked:
                note = f"Earnings window: {business_days_until} business day(s) until expected report"
            elif next_date:
                note = f"Expected report in {business_days_until} business day(s)"
            else:
                note = f"No expected report in the next {LOOKAHEAD_DAYS} calendar days"
        risks[symbol] = {
            "symbol": symbol,
            "status": status,
            "nextDate": next_date,
            "businessDaysUntil": business_days_until,
            "source": "NASDAQ",
            "note": note,
        }
    return risks
